File: DrawingWithNoise/src/DrawingWithNoise.java
// Main Class
// Drawing With Noise
// Started: 6 August 2019
// Press 'a' to change background color
// Press 's' to randomly change color type
// Press 'd' to turn glow effect on and off

import java.util.ArrayList;
import processing.core.PApplet;


public class DrawingWithNoise extends PApplet{

    private ArrayList<Line> lines;
    private ArrayList<Point> points;
    private float meshSize;
    private float n;
    private Color color;
    private boolean glow;
    private float timer;
    private float timerMax;
    private boolean blackBG;

    public static void main(String[] passedArgs) {
        String[] appletArgs = new String[] { "DrawingWithNoise" };  // first string MUST match name of class

        if (passedArgs != null) {
            PApplet.main(concat(appletArgs, passedArgs));
        } else {
            PApplet.main(appletArgs);
        }

    }

    public void settings() {
        fullScreen(P2D);
    }

    public void setup() {
        lines = new ArrayList<>();
        points = new ArrayList<>();
        meshSize = 50;
        n = 0;
        color = new Color(this);
        glow = false;
        timer = 0;
        timerMax = 500;
        blackBG = true;
    }

    public void draw() {

        if (blackBG) {
            background(0);
        } else {
            background(255);
        }

        if (glow) {
            blendMode(ADD);
        } else {
            blendMode(BLEND);
        }

        float x = random(mouseX - meshSize, mouseX + meshSize);
        float y = random(mouseY - meshSize, mouseY + meshSize);
        Point p = new Point(this, x, y);
        int col = getColor();
        p.setColor(col);
        points.add(p);
        addLines();

        for (Line line: lines) {
            line.display();
        }

        for (Point point: points) {
            point.display();
        }

        removeFadedLines();
        removeFadedPoints();
        meshSize = noise(n) * min(width, height) / 8;
        n = (n + 0.01f) % 50000f;
        timer = (timer + 0.1f) % timerMax;
    }

    private void removeFadedLines() {
        ArrayList<Line> toRemove = new ArrayList<>();

        for (Line line: lines) {

            if (line.isFaded()) {
                toRemove.add(line);
            }
        }

        for (Line line: toRemove) {
            lines.remove(line);
        }
    }

    private void removeFadedPoints() {
        ArrayList<Point> toRemove = new ArrayList<>();

        for (Point point: points) {

            if (point.isFaded()) {
                toRemove.add(point);
            }
        }

        for (Point point: toRemove) {
            points.remove(point);
        }
    }

    private void addLines() {
        int lastIndex = points.size() - 1;
        Point lastPoint = points.get(lastIndex);

        for (int i = 0; i < lastIndex; i++) {
            Point point = points.get(i);

            if (point.getAlpha() > 200) {

                if (!isPointIntersectingCircle(lastPoint, point)) {
                    Line line = new Line(this, lastPoint, point);
                    int c = getColor();
                    line.setColor(c);
                    lines.add(line);
                }
            }
        }
    }

    private boolean isPointIntersectingCircle(Point a, Point b) {
        boolean isIntersectingPoint = false;
        float diameter = a.distance(b);
        Point center = createCenterPoint(a, b);

        for (Point compare: points) {

            if (compare != a && compare != b && compare.getAlpha() > 200) {
                float distance = compare.distance(center);

                if (distance < diameter / 2) {
                    isIntersectingPoint = true;
                    break;
                }
            }
        }

        return  isIntersectingPoint;
    }

    private Point createCenterPoint(Point a, Point b) {
        float centerX = (a.getPosition().x + b.getPosition().x) / 2;
        float centerY = (a.getPosition().y + b.getPosition().y) / 2;
        Point center = new Point(this, centerX, centerY);
        return center;
    }

    public void keyPressed() {

        if (key == 'a') {
            blackBG = !blackBG;

            if (!blackBG) {
                glow = false;
            }

        } else if (key == 's') {
            color.setRandomColorType();

        } else if (key == 'd') {

            if (blackBG) {
                glow = !glow;
            }
        }
    }

    public int getColor() {
        int col = color(255);

        if (color.isRGB()) {
            colorMode(RGB, 255);
            col = color.randomColor();
        } else if (color.isHSB()) {
            colorMode(HSB, 360);
            col = color.mapColor(timer, 0, timerMax, 360, 360);
        }

        return col;
    }
}

File: docs/DrawingWithNoise/drawing_with_noise.js
// Drawing with Noise sketch
// Press 'a' to change background color
// Press 's' to change color scheme

var points;
var lines;
var meshSize;
var n;
var deltaAlpha;
var hasBlackBackground;

function setup() {
    createCanvas(windowWidth-20, windowHeight-20, P2D);
    points = [];
    lines = [];
    meshSize = 50;
    n = 0;
    deltaAlpha = -1;
    hasBlackBackground = true;
    frameRate(60);
}

function draw() {
    displayBackground();
    addPoint();
    addLines();
    displayPoints();
    displayLines();
    removeFadedPoints();
    removeFadedLines();
    setMeshSize();
}

function keyTyped() {

    if (key == 'a') {
        hasBlackBackground = !hasBlackBackground;
    }
}

function displayBackground() {

    if (hasBlackBackground) {
        background(0);
    } else {
        background(255);
    }
}

function addPoint() {
    let x = random(mouseX - meshSize, mouseX + meshSize);
    let y = random(mouseY - meshSize, mouseY + meshSize);
    points.push(new Point(x, y));
}

function addLines() {
    let lastIndex = points.length - 1;
    let lastPoint = points[lastIndex];

    for (let i = 0; i < lastIndex; i++) {
        let point = points[i];

        if (point.getAlpha() > 200) {
            if (!hasPointIntersectingCircle(lastPoint, point)) {
                lines.push(new Line(lastPoint, point));
            }
        }
    }
}

function hasPointIntersectingCircle(a, b) {
    let hasIntersectingPoint = false;
    let diameter = a.distance(b);
    let center = createCenterPoint(a, b);

    for (let i = 0; i < points.length; i++) {
        let point = points[i];

        if (!Object.is(point, a) && !Object.is(point, b) && point.getAlpha() > 200) {
            let distance = point.distance(center);

            if (distance < diameter / 2) {
                hasIntersectingPoint = true;
                break;
            }
        }
    }

    return hasIntersectingPoint;
}

function createCenterPoint(a, b) {
    let centerX = (a.position.x + b.position.x) / 2;
    let centerY = (a.position.y + b.position.y) / 2;
    return (new Point(centerX, centerY));
}

function displayPoints() {

    for (let i = 0; i < points.length; i++) {
        points[i].display();
    }
}

function displayLines() {

    for (let i = 0; i < lines.length; i++) {
        lines[i].display();
    }
}

function removeFadedPoints() {
    for (let i = points.length - 1; i >= 0; i--) {
        let point = points[i];

        if (point.isFaded()) {
            points.splice(i, 1);
        }
    }
}

function removeFadedLines() {
    for (let i = lines.length - 1; i >= 0; i--) {
        let line = lines[i];

        if (line.isFaded()) {
            lines.splice(i, 1);
        }
    }
}

function setMeshSize() {
    meshSize = noise(n) * min(width, height) / 8;
    n += 0.01;

    if (n > 50000) {
        n = 0;
    }
}
